File: haskell/Project Euler/29_uniqueApowerB.hs
import IO
import List

sieve :: [Int] -> [Int]
sieve [] = []
sieve (x : xs) = x : sieve [p | p <- xs, p `mod` x /= 0]

primes n = (2 : sieve [3, 5 .. n])

primes100 = primes 100

prime_divisors_and_powers :: Int -> [(Int, Int)]
prime_divisors_and_powers n = 
	sanitized_divisors
	where
		primesList = filter (<=n) primes100
		divisors = prime_divisors_and_powers' n primesList []
		sanitized_divisors = filter (\x -> (snd(x) > 0)) divisors
	
prime_divisors_and_powers' n [] divisors = divisors
prime_divisors_and_powers' n [p] divisors = 
		divisors ++ [(p, nPowers)]
		where
			nPowers = (snd (divisor_power n p 0))
prime_divisors_and_powers' n (px:pxs) divisors = 
	if pxs == []
		then divisors
		else
			prime_divisors_and_powers' new_n pxs (divisors ++ [(px, nPowers)])
			where
				(new_n, nPowers) = divisor_power n px 0

divisor_power n p nPowers = 
	if n `mod` p /= 0
		then (n, nPowers)
		else
			divisor_power (n `div` p) p (nPowers + 1)

-- uniqueApowerB :: [Int] -> [Int] -> [[(Int, Int)]]
uniqueApowerB alist blist = 
	nub (foldl (\x y -> x ++ (curr_exp y)) [] blist)
	where
	adivisors = map (\x -> (prime_divisors_and_powers x)) alist
	add_powers y divisors = map (\x -> (fst(x), (snd(x) * y))) divisors
	curr_exp y = map (\x -> add_powers y x) adivisors

main = do
	print $ length (uniqueApowerB [2..5] [2..5]) {- 15 -}
	print $ length (uniqueApowerB [2..100] [2..100]) {- 9183 -}

File: python/project_euler/29_uniqueApowerB.py
#!/usr/bin/python

from Primes1mil import *

def prime_divisors_and_powers (n):
	i = 0

	divisors = []
	while (n > 1):
		nPowers = 0
		# print primes[i], n

		while ( n % primes[i] == 0):
			nPowers += 1
			n /= primes[i]

		if (nPowers > 0):
			divisors.append((primes[i], nPowers))
		i += 1

	return divisors

def valuesFromDivisors (div):
	return (reduce( (lambda x, y: x + (y[0] ** y[1])), div, 0))

def uniqueApowerB (a, b):
	uniq_values = []

	for x in a:
		divisors = prime_divisors_and_powers(x)
		for y in b:
			curr_exp = map( (lambda l: (l[0], (l[1]*y))), divisors)

			if curr_exp not in uniq_values:
				uniq_values.append(curr_exp)
	return uniq_values

def main():
	lst = uniqueApowerB(range(2, 101), range(2, 101))
	print len(lst)
	# 9183

if __name__ == "__main__":
	main()

